src/imputation/MoR.py

```diff
@@ -0,0 +1,98 @@
+"""Functions for the Mean of Ratios (MoR) methods."""
+import pandas as pd
+import re
+
+from src.imputation.apportionment import run_apportionment
+
+good_statuses = ["Clear", "Clear - overridden"]
+bad_statuses = ["Form sent out", "Check needed"]
+
+
+def run_mor(df, backdata, target_vars):
+    """Function to implement Mean of Ratios method.
+
+    This is implemented by first carrying forward data from last year
+    for non-responders, and then calculating and applying growth rates
+    for each imputation class.
+
+    Args:
+        df (pd.DataFrame): Processed full responses DataFrame
+        backdata (pd.DataFrame): One period of backdata.
+        target_vars ([string]): List of variables to impute.
+
+    Returns:
+        pd.DataFrame: df with MoR applied
+    """
+    to_impute_df, remainder_df, backdata = mor_preprocessing(df, backdata)
+
+    # Carry forwards method
+    carried_forwards_df = carry_forwards(to_impute_df, backdata, target_vars)
+
+    # TODO Remove the `XXX_prev` columns (left in for QA)
+    return pd.concat([remainder_df, carried_forwards_df])
+
+
+def mor_preprocessing(df, backdata):
+    """Apply pre-processing ready for MoR
+
+    Args:
+        df (pd.DataFrame): full responses for the current year
+        backdata (pd.Dataframe): backdata file read in during staging.
+    """
+    # TODO move this to imputation main
+    # Select only values to be imputed and remove duplicate instances
+    imputation_condition = (
+        (df["formtype"] == "0001")
+        & (df["status"].isin(bad_statuses))
+        & (df["instance"] == 0 | pd.isnull(df["instance"]))
```

Comment:
for instance 0 there are no values for the target variables, instance 0 holds metadata and in long forms the target variables appear on the other instances. Also, instance 0 doesn't have "C" and "D" values so the don't belong to an imputation class. Could you work with instance 1 instead?

Reply:
That might be a good idea - we're using this to pull out the bad statuses in the data which we then go on to overwrite some of using carry forwards. However at the moment this does mean we'd lose stuff going into TMI.

Reply:
I think that we want to use MoR as the first priority, so we want to do as much as possible with MoR instead of TMI (which isn't supposed to be as good). So we'd want to pull out all the baddies and do as many as possible with MoR Does that make sense with how you're thinking of it?

Reply:
Yeh I think so - the reason I filter by instance is that we only have unique references in the current data when we join to the backdata. That way the pattern of instance is carried over from the backdata. In terms of the metadata in instance one does that make sense?

Reply:
Ah yes, so for carried forward you're not using imputation classes, so of course, just keeping instance 0 makes sense.

Reply:
Although when we come to add MoR we will need imputation classes so don't know if that means we should use instance 1?